cabal-install/src/Distribution/Client/InstallPlan.hs

```diff
@@ -170,16 +172,22 @@ import qualified Distribution.Compat.Graph as Graph
 -- dependencies; if we give a 'PackageInstalled' instance it would be too easy
 -- to get this wrong (and, for instance, call graph traversal functions from
 -- Cabal rather than from cabal-install). Instead, see 'PackageInstalled'.
+--
+-- We additionally attach to the packages in the install plan the constraint
+-- scope in which they were chosen (because, in a context with qualified goals,
+-- there may be multiple versions of the same package in an install plan, and
+-- e.g. in `cabal freeze` we need to know in which scope was solved any given
+-- package)
 data GenericPlanPackage ipkg srcpkg
-  = PreExisting ipkg
-  | Configured srcpkg
-  | Installed srcpkg
+  = PreExisting ipkg ConstraintScope
+  | Configured srcpkg ConstraintScope
+  | Installed srcpkg ConstraintScope
```

Comment:
In cabal-install our understanding of the install plan must also consider, for every package in the plan, the scope it was solved in.